src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -73,58 +83,131 @@ public void storeMessage(Message message) {
-    Filter directMessages = CompositeFilterOperator.or(loggedInUserMessages, otherUserMessages);
+    return CompositeFilterOperator.or(loggedInUserMessages, otherUserMessages);
+  }
 
-    // Messages are ordered from oldest to newest since messages between
-    // people may include implicit context/reference to previously sent messages,
-    // and this order makes the chronology of the conversation and context
-    // more intuitive and understandable in the messages list page. 
-    Query query =
-        new Query("Message")
-            .setFilter(directMessages)
-            .addSort("timestamp", SortDirection.ASCENDING);
+  public List<Message> getAllMessages() {
+    Query query = createPublicMessageQuery();
     PreparedQuery results = datastore.prepare(query);
+    return prepareQueryToMesssages(results);
+  }
 
-    for (Entity entity : results.asIterable()) {
-      String idString = entity.getKey().getName();
-      UUID id = UUID.fromString(idString);
-      String text = getStringProperty(entity, "text").orElse("");
-      long timestamp = (long) entity.getProperty("timestamp");
-      String sender = (String) entity.getProperty("user");
-      String receiver = (String) entity.getProperty("recipient");
-      Message message = new Message(id, sender, text, timestamp, receiver);
-      messages.add(message);
+  private List<Message> prepareQueryToMesssages(PreparedQuery query) {
+    List<Message> results = new ArrayList<>();
+    for (Entity entity : query.asIterable()) {
+      results.add(entityToMessage(entity));
     }
+    return results;
+  }
 
-    return messages;
+  private Message entityToMessage(Entity entity) {
+    String idString = entity.getKey().getName();
+    UUID id = UUID.fromString(idString);
+    String text = getStringProperty(entity, "text").orElse("");
+    long timestamp = (long) entity.getProperty("timestamp");
+    String sender = (String) entity.getProperty("user");
+    String receiver = (String) entity.getProperty("recipient");
+    return new Message(id, sender, text, timestamp, receiver);
+  }
+
+  private Query createPublicMessageQuery() {
+    return new Query("Message").addSort("timestamp", SortDirection.ASCENDING);
   }
 
-  /** Returns the total number of messages for all users. */
   public int getTotalMessageCount() {
     Query query = new Query("Message");
     PreparedQuery results = datastore.prepare(query);
     return results.countEntities(FetchOptions.Builder.withDefaults());
   }
 
+  /**
+   * This is to create an user form data to put in the datastore Going with option 2 highlighted in
+   * the Meeting Notes
+   *
+   * @throws EntityNotFoundException
+   */
+  public void storeUserTeaData(
+     Map<String, Long> incomingUserTeaData, String username, String date) {
+
+    Entity datastoreUser;
+    String keyName = date + ":" + username;
+    Key usernameKey = KeyFactory.createKey("UserTeaData", keyName);
+
+    try {
+      datastoreUser = datastore.get(usernameKey);
```

Comment:
When it comes to `try`...`catch`, it is best to minimize the amount of content within the `try` block. When there are too many statements in a single `try` block, it increases the number of ways that the try might fail resulting in an exception being caught in the corresponding handler code, and it makes it confusing to other readers regarding which of the various possible sources of error you were intending to handle.

I would recommend simplifying it like this:

    public void storeUserTeaData(
        String username,
        LocalDate date,
        ZoneId timeZone,
        Map<String, Integer> histogram) {
       Key key = createTeaUserDataKey(username, date, timeZone);
       Entity entity = null;
       try {
         entity = datastore.get(key);
       } catch (EntityNotFoundException e) {
         entity = createNewUserTeaDataEntity(key, username, date, timeZone);
       }
       setUserTeaDataHistogram(histogram, entity);
       datastore.put(entity);
    }

That is, your `try` should only wrap the `get` operation. Additionally, by creating subroutines here, you can minimize repetition; for example, the code to populate the histogram into the entity is shared on both code paths and should not need to be repeated.